app/master/build_request_handler.py

```diff
@@ -102,6 +102,10 @@ def _prepare_build_async(self, build, project_lock):
             try:
                 build.prepare(self._subjob_calculator)
                 if not build.has_error:
+                    # If the atomizer found no work to do, perform build cleanup and skip the slave allocation.
+                    if len(build.all_subjobs()) == 0:
+                        self._logger.info('Build {} has no work to perform and is exiting.', build.build_id())
+                        build.finish()
```

Comment:
I know you already merged but I just noticed this -- should this return after calling finish()?